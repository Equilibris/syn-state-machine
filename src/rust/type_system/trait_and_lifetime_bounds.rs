use crate::*;

pub type TypeParamBounds<Attr, Ty> = MinLength<InterlaceTrail<TypeParamBound<Attr, Ty>, Plus>>;

#[derive(Debug)]
pub enum TypeParamBound<Attr, Ty> {
    Lt(Lifetime),
    Tr(TraitBound<Attr, Ty>),
}
impl<Attr: Parse, Ty: Parse> Parse for TypeParamBound<Attr, Ty> {
    fn parse<'a>(input: &mut ParseBuffer<'a>) -> Result<Self> {
        Ok(match input.parse::<Sum2<_, _>>()? {
            Sum2::V0(a) => Self::Lt(a),
            Sum2::V1(a) => Self::Tr(a),
        })
    }
}

materialize! {
    #[derive(Debug)]
    pub struct TraitBound<Attr, Ty> {
        q peek <- Question;
        for_lts <- Option<ForLifetimes<Attr, Ty>>;
        path <- TypePath<Ty>;
    }
}

pub type LifetimeBounds = MinLength<Interlace<Lifetime, Plus>>;
pub type Lifetime = LifetimeToken;

materialize! {
    #[derive(Debug)]
    pub struct ForLifetimes<Attr, Ty> {
        <- KwFor;
        args <- GenericParams<Attr, Ty>;
    }
}
