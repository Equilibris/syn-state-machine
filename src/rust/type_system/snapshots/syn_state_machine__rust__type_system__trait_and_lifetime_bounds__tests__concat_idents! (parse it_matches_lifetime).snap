---
source: src/rust/type_system/trait_and_lifetime_bounds.rs
expression: "crate::parse::<Lifetime>(q.clone())"
---
Ok(
    LifetimeToken {
        ident: Ident(
            a,
        ),
    },
)
